storage/src/test/java/org/apache/kafka/storage/internals/log/LogSegmentTest.java

```diff
@@ -0,0 +1,819 @@
+            FetchDataInfo read = seg.read(55, 200);
+            checkEquals(ms2.records().iterator(), read.records.records().iterator());
+        }
+    }
+
+    @ParameterizedTest(name = "testReadWhenNoMaxPosition minOneMessage = {0}")
+    @ValueSource(booleans = {true, false})
+    public void testReadWhenNoMaxPosition(boolean minOneMessage) throws IOException {
+        Optional<Long> maxPosition = Optional.empty();
+        int maxSize = 1;
+        try (LogSegment seg = createSegment(40)) {
+            MemoryRecords ms = records(50, "hello", "there");
+            seg.append(51, RecordBatch.NO_TIMESTAMP, -1L, ms);
+
+            // read before first offset
+            FetchDataInfo read = seg.read(48, maxSize, maxPosition, minOneMessage);
+            assertEquals(new LogOffsetMetadata(48, 40, 0), read.fetchOffsetMetadata);
+            assertFalse(read.records.records().iterator().hasNext());
+
+            // read at first offset
+            read = seg.read(50, maxSize, maxPosition, minOneMessage);
+            assertEquals(new LogOffsetMetadata(50, 40, 0), read.fetchOffsetMetadata);
+            assertFalse(read.records.records().iterator().hasNext());
+
+            // read at last offset
+            read = seg.read(51, maxSize, maxPosition, minOneMessage);
+            assertEquals(new LogOffsetMetadata(51, 40, 39), read.fetchOffsetMetadata);
+            assertFalse(read.records.records().iterator().hasNext());
+
+            // read at log-end-offset
+            read = seg.read(52, maxSize, maxPosition, minOneMessage);
+            assertNull(read);
+
+            // read beyond log-end-offset
+            read = seg.read(53, maxSize, maxPosition, minOneMessage);
+            assertNull(read);
+        }
+    }
+
+    /**
+     * In a loop append two messages then truncate off the second of those messages and check that we can read
+     * the first but not the second message.
+     */
+    @Test
+    public void testTruncate() throws IOException {
+        try (LogSegment seg = createSegment(40)) {
+            long offset = 40;
+            for (int i = 0; i < 30; i++) {
+                MemoryRecords ms1 = records(offset, "hello");
+                seg.append(offset, RecordBatch.NO_TIMESTAMP, -1L, ms1);
+                MemoryRecords ms2 = records(offset + 1, "hello");
+                seg.append(offset + 1, RecordBatch.NO_TIMESTAMP, -1L, ms2);
+
+                // check that we can read back both messages
+                FetchDataInfo read = seg.read(offset, 10000);
+                assertEquals(Arrays.asList(ms1.records().iterator().next(), ms2.records().iterator().next()), iteratorToList(read.records.records().iterator()));
+
+                // Now truncate off the last message
+                seg.truncateTo(offset + 1);
+                FetchDataInfo read2 = seg.read(offset, 10000);
+                assertEquals(1, iteratorToList(read2.records.records().iterator()).size());
```

Comment:
all we want to verify is the content of records, so we can use `assertIterableEquals(ms1.records(), read2.records.records());` instead